### PR TITLE
feat(torchx): add batching support for dot

### PR DESCRIPTION
**Pull request by polvalente**
helps with #682

### CODE REVIEW

File: torchx/lib/torchx/backend.ex [hunk 1]
```diff
@@ -921,6 +921,29 @@ defmodule Torchx.Backend do
...
+  def dot(
+        %T{type: out_type} = out,
+        %T{type: left_type, data: %TB{ref: left_ref}},
+        left_axes,
+        left_batched_axes,
+        %T{type: right_type, data: %TB{ref: right_ref}},
+        right_axes,
+        right_batched_axes
+      ) do
+    left_axes = Enum.map(left_axes, fn x -> x - Enum.count(left_batched_axes, &(&1 < x)) end)
```
Comment (polvalente) [inline torchx/lib/torchx/backend.ex:933]: @josevalim @jonatanklosko any idea on how to make this more efficient? Basically, I need to "remove" the batching axes from the axis numbering for libtorch

File: torchx/lib/torchx/backend.ex [hunk 2]
```diff
@@ -921,6 +921,29 @@ defmodule Torchx.Backend do
...
+    right_axes = Enum.map(right_axes, fn x -> x - Enum.count(right_batched_axes, &(&1 < x)) end)
+
+    Torchx.batched_tensordot(
```
Comment (polvalente) [inline torchx/lib/torchx/backend.ex:936]: Should I try to make tensordot optionally aware of batching instead of exposing a separate function?

Comment (josevalim) [reply torchx/lib/torchx/backend.ex:936]: I think it is fine as is, I would just remove the `batched_` prefix.

File: torchx/lib/torchx/backend.ex (hunk 1 above)
Comment (josevalim) [reply torchx/lib/torchx/backend.ex:933]: You could do manual recursion to avoid multiple traversals but it probably doesn't matter.

**Review by josevalim — APPROVED** [review]

File: torchx/test/torchx/nx_test.exs [hunk 3]
```diff
@@ -420,18 +420,6 @@ defmodule Torchx.NxTest do
       )
     end
 
-    test "dot currently raises when using batching" do
```
Comment (Ian-GL) [inline torchx/test/torchx/nx_test.exs:423]: The rest of the dot related tests here could also be removed as they are testing the same thing as the doctests, so now that those are restored they are not needed.

File: torchx/lib/torchx/backend.ex (hunk 2 above)
Comment (polvalente) [reply torchx/lib/torchx/backend.ex:936]: I ended up merging all into a single tensordot/6 clause for the backend and a single function at the cpp level

**Pull request merged**